actions/run-nrpe-check

```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-check=$(action-get name | sed -e 's/-/_/g')
+check=$(action-get name)
```

Comment:
I'm wondering why this name substituting behaviour was originally added - will it break anything else if we remove it?

Reply:
I have wondered the same myself as I was working on the bugfix. My theory is that, because the result is passed to the juju action result in "action-set" as a key:value structure where the key cannot have "_", it motived this whole replacement of "-" and "_", so when listing you had to remove the "_" otherwise you couldn't list, and then when running, you had to add it back because the user would have seen the name with "-" instead of "_", however, whoever designed this didn't take into account that some filenames may have both "_" and "-" in them, so you never get back the original file name when you perform those string replacements. The solution I  found was to overcome the juju action result structure limitation by passing a whole string as the value in the result instead of key:value structure per line.

Reply:
Thanks, that makes sense. The only concern I have is that this is a breaking change for the interface of list-nrpe-checks. I'm not sure if anything would be programmatically using the output of list-nrpe-checks though?

---

For reference, this is the difference in output:

```
❯ juju run nrpe-current/0 list-nrpe-checks
Running operation 63 with 1 task
  - task 64 on unit-nrpe-0

Waiting for task 64...
checks:
  check-conntrack: /usr/local/lib/nagios/plugins/check_conntrack.sh -w 80 -c 90
  check-reboot: /usr/local/lib/nagios/plugins/check_reboot.py "2024-12-03 21:22:42"
  check-systemd-scopes: /usr/local/lib/nagios/plugins/check_systemd_scopes.py -e 25
    -E 50 -a 25 -A 25
timestamp: Tue Dec  3 22:54:18 UTC 2024

❯ juju run nrpe-patched/0 list-nrpe-checks
Running operation 65 with 1 task
  - task 66 on unit-nrpe-patched-0

Waiting for task 66...
checks: |
  check_conntrack: /usr/local/lib/nagios/plugins/check_conntrack.sh -w 80 -c 90
  check_reboot: /usr/local/lib/nagios/plugins/check_reboot.py "2024-12-03 22:34:49"
  check_systemd_scopes: /usr/local/lib/nagios/plugins/check_systemd_scopes.py -e 25 -E 50 -a 25 -A 25
timestamp: Tue Dec  3 22:54:21 UTC 2024
```

For human operators, this is pretty similar, but if anyone is using this in a script to retrieve and run checks, it will be affected. Eg.

```
❯ juju run nrpe/0 list-nrpe-checks --format json | jq '.[].results.checks'
Running operation 85 with 1 task
  - task 86 on unit-nrpe-0

Waiting for task 86...
{
  "check-conntrack": "/usr/local/lib/nagios/plugins/check_conntrack.sh -w 80 -c 90",
  "check-reboot": "/usr/local/lib/nagios/plugins/check_reboot.py \"2024-12-03 21:22:42\"",
  "check-systemd-scopes": "/usr/local/lib/nagios/plugins/check_systemd_scopes.py -e 25 -E 50 -a 25 -A 25"
}

❯ juju run nrpe-patched/0 list-nrpe-checks --format json | jq '.[].results.checks'
Running operation 87 with 1 task
  - task 88 on unit-nrpe-patched-0

Waiting for task 88...
"check_conntrack: /usr/local/lib/nagios/plugins/check_conntrack.sh -w 80 -c 90\ncheck_reboot: /usr/local/lib/nagios/plugins/check_reboot.py \"2024-12-03 22:34:49\"\ncheck_systemd_scopes: /usr/local/lib/nagios/plugins/check_systemd_scopes.py -e 25 -E 50 -a 25 -A 25\n"
```

Reply:
I think this is a breaking change based on the output.

Would it be possible to compare the file names before executing? For example, we could replace all underscores (_) with dashes (-) in the input target file and the local scrape files, then compare them to find the matching target.

Reply:
Here is the bug and PR on launchpad

Bug: https://bugs.launchpad.net/charm-nrpe/+bug/2017602
PR: https://code.launchpad.net/~raychan96/charm-nrpe/+git/charm-nrpe/+merge/443187

It's better to validate the orignal issue before merge this one.

Reply:
@Pjack that bug is unrelated to this one, and the Ray's patch you linked to was merged and is already in this code base.

Reply:
@samuelallan72 : I agree that it is a script-breaking change and non-backwards compatible, but between this solution and the problem, I'd choose the solution here. Scripts could be updated. However @jneo8 's idea is another possible solution, although I find it a bit controversial. It cannot be implemented as stated, because files have both _ and - in them. What we could do is create a regex replacing every _ and - by . and see what matches (in other words, treat both as the same). That way we don't have to change listing, and running will find the file. However, programatically speaking, this is sort of imprecise programming, if there are ever 2 files with the same name but different combination of _ and -, it will break the logic or one of them will never get run. Such situation is very unlikely to happen though.

Reply:
Implemented @jneo8 suggestion

Reply:
It's not perfect. But it's also not make sense if there are two NRPE files with similar name.

Thanks for the implementation!

Reply:
What are the chances that a file name will be an invalid regex? It should be low I think. :thinking:

Reply:
invalid regex would be impossible. The input the user types in is the name of the check, so the user would have typed it wrong, or there is no matching file. The regex covers all variations of - and _ so it doesn't matter if the user types in - or _ in the name, as long as there is a file with either (or both).

Reply:
If the regex is based on the filename, it could be an invalid regex - for example `check((.py`. Very unlikely I guess, but still possible.

Reply:
Resolving the conversation, because the concerns about breaking changes have been addressed. I think the new matching method is good enough here. :+1: